File: retired/hard/intentions/fileRecovery.py
#!/usr/bin/env python3
import os, subprocess  # Import the necessary libraries.
import hashlib
import argparse

# Create an ArgumentParser object to handle script arguments.
parser = argparse.ArgumentParser(description='Recover file using /opt/scanner/scanner.')
# Add arguments for file to recover and output file. These are required arguments.
parser.add_argument('-f', '--file', required=True, help='The file to recover.')
parser.add_argument('-o', '--output', required=True, help='The output file.')
args = parser.parse_args()  # Parse the input arguments.

filename = args.file  # File to recover.
output = args.output  # Output file name.

# Define a function to scan the file using an external scanner command.
def scan_file(num_bytes):
    # Prepare the scanner command.
    command = f"/opt/scanner/scanner -c {filename} -p -l {num_bytes} -s a".split()
    # Run the command and wait for it to complete.
    popen = subprocess.Popen(command, stdout=subprocess.PIPE)
    popen.wait()
    # Get the command output.
    output = popen.stdout.read().decode("utf-8")
    # Extract the hash value from the output.
    hash = output.split(" ")[-1]
    return hash

# Main execution starts here.
i = 1  # Initialize counter.
file_so_far = bytearray()  # Initialize the byte array to hold the file content.

# Start a loop to continuously attempt recovery.
while True:
    hash = scan_file(i).strip()  # Get the hash of the current chunk of file.
    found = False  # Flag to indicate if a match is found.

    # Try each possible byte value.
    for j in range(0x01, 0xff):
        byte = bytes([j])  # Convert integer to byte.
        attempt = file_so_far.copy()  # Make a copy of the current file content.

        # Append the current byte to the attempt.
        attempt += bytearray(byte)
        # Calculate the MD5 hash of the attempt.
        new_hash = hashlib.md5(attempt).hexdigest()

        # If the hashes match, update the file content and break the loop.
        if new_hash == hash:
            file_so_far = attempt
            found = True
            break

    # If no matching byte is found, print an error and break the main loop.
    if not found:
        print("ERROR: could not find valid byte")
        break

    i += 1  # Increment the counter.

# Write the recovered file content to the output file.
with open(output, "wb") as f:
    f.write(file_so_far)

File: retired/hard/intentions/fileRecovery.pl
#!/usr/bin/perl
use strict;  # Enable strict mode to catch possible errors.
use warnings;  # Enable warnings to catch potential problems.
use Digest::MD5 qw(md5_hex);  # Use MD5 hashing algorithm.
use Getopt::Long qw(GetOptions);  # Use GetOptions for command line arguments.

# Declare variables for file to recover and output file.
my $filename;
my $output;

# Handle command line arguments with GetOptions.
GetOptions(
    'file|f=s' => \$filename,  # Get file name to recover.
    'output|o=s' => \$output   # Get name of output file.
) or die "Usage: $0 --file FILE --output OUTPUT\n";  # If arguments are not provided, print usage and exit.

# Define function to scan file.
sub scan_file {
    my ($num_bytes) = @_;  # Get the number of bytes to read.
    # Prepare the scanner command.
    my $command = "/opt/scanner/scanner -c $filename -p -l $num_bytes -s a";
    # Execute the command and capture its output.
    my $output = `$command`;
    # Split the output into words.
    my @words = split ' ', $output;
    my $hash = $words[-1];    # Get the last word (the hash).
    return $hash;  # Return the hash.
}

# Main program starts here.
my $i = 1;  # Initialize counter.
my $file_so_far = "";  # Initialize the file content.

# Start a loop to continuously attempt recovery.
while (1) {
    my $hash = scan_file($i);  # Get the hash of the current file chunk.
    $hash =~ s/\s+$//;   # Remove trailing spaces.
    my $found = 0;  # Flag to indicate if a match is found.

    # Try each possible byte value.
    for my $j (0x01..0xff) {
        my $byte = pack('C', $j);  # Convert integer to byte.
        my $attempt = $file_so_far . $byte;  # Append the byte to the file content.
        my $new_hash = md5_hex($attempt);  # Compute the MD5 hash of the attempt.

        # If the hashes match, update the file content and break the loop.
        if ($new_hash eq $hash) {
            $file_so_far = $attempt;
            $found = 1;
            last;
        }
    }

    # If no matching byte is found, print an error and break the main loop.
    if (!$found) {
        print("ERROR: could not find valid byte\n");
        last;
    }

    $i += 1;  # Increment the counter.
}

# Write the recovered file content to the output file.
open my $fh, '>', $output or die "Could not open file '$output' $!";  # Open the output file.
print $fh $file_so_far;  # Write the file content.
close $fh;  # Close the file.
